### PR TITLE
Change API type `Tensor?[]` for structured kernels.

### PR DESCRIPTION
**Pull request by ysiraichi**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #73351

Partially fixes: #66328

(basically, same as previous PR, but for `IOptTensorRefList`)

This PR:
- adds support for `IOptTensorRefList` to the dispatcher for:
  - computing the dispatch key
  - boxing and unboxing `IOptTensorRefList`
- modified the codegen for structured kernels:
  - codegen APIs use `IOptTensorRefList` instead of `List<optional<Tensor>>`

**Changes summary:**

- Miscelaneous functions used by codegen'd functions (e.g. `FunctionalTensorWrapper.*`)
- Dispatcher changes for handling `IOptTensorRefList` correctly (e.g. `DispatchKeyExtractor.h`)
- Forward declarations of `IOptTensorRefList` (e.g. `MethodOperators.h`)
- Signature changes of `at::index` due to the need of `const` inside `TensorBody.h`
- Codegen changes, special casing structured kernels (e.g. `gen_variable_type.py`)

cc @EikanWang @jgong5 @wenzhe-nrv @sanchitintel @ezyang @bhosmer @bdhirsh

### CODE REVIEW

File: tools/autograd/gen_variable_type.py [hunk 1]
```diff
@@ -218,15 +228,15 @@
 SAVE_OPTIONALTENSORLIST_IMPL = CodeTemplate("""\
 std::vector<c10::intrusive_ptr<TensorImpl>> ${tensorlist_name}_impl_saved(${tensorlist_name}.size());
 for (size_t i=0; i<${tensorlist_name}.size(); i++) {
-  c10::optional<Tensor> t = ${tensorlist_name}[i];
-  if (t.has_value() && t->defined()) ${tensorlist_name}_impl_saved[i] = t->getIntrusivePtr();
+  auto& t = ${tensorlist_name}[i];
+  if (t.has_value()) ${tensorlist_name}_impl_saved[i] = t->getIntrusivePtr();
```
Comment (albanD) [inline tools/autograd/gen_variable_type.py:232]: Is it always ok to remove the defined check here?

Comment (ysiraichi) [reply tools/autograd/gen_variable_type.py:232]: Ah, nice catch! It's not ok (it was when only `OptionalTensorRef` got to this point).

**Comment by ysiraichi** [comment]
@ezyang @bdhirsh 
Apparently, some lazy tensor tests are failing. I've written about it below, so let me know what you think.

-----

### The Cause
- Call: `lazy_self[lazy_index]`
- Gets dispatched to the `CompositeExplicitAutograd` version of `index.Tensor`
- Gets dispatched to the `ltc_eager_fallback`, which copies the `lazy_self` tensor (only) to CPU
- Assert fail: can't have `lazy_self` on CPU and `lazy_indices` on Lazy device

-----

### Why does it fail?
- Only `lazy_self` gets copied to CPU
- Since `lazy_index` is passed as a list of **optional** tensors, it doesn't get copied to CPU
- `ts_eager_fallback` (called by `ltc_eager_fallback`) does not handle lists of optional tensors

-----

### Possible Solutions
1. Make `ts_eager_fallback` handle list of optional tensors (sounds better)
2. Remove `index.Tensor` testing for lazy tensors

**Comment by ysiraichi** [comment]
@ezyang @bdhirsh 
Here's a summary of the recent changes:
- Similar to the changes in the previous PR:
    - Fixed the JIT type construction for `OptionalTensorRef` (same thing)
    - Change `tensorlist_has_dispatch` argument type
    - ...
- Changed some of the lazy tests: `TestMultiIndexHeadNull`, `TestMultiIndexMiddleNull`, and `TestMultiIndexTailNull`.
    - **Problem:** `OptionalTensorRef` cannot be created from undefined tensors. So, instead of passing an undefined tensor, I had to braket-initialize
    - I think this is BC-breaking, so maybe we should fix it before merging?
    - **Suggestion:** instead of having `IOptTensorListRef` as the C++ API, change it to `vector<Tensor>` (assuming that undefined tensors are `nullopt`)

**Comment by ysiraichi** [comment]
@bdhirsh It seems that the only failing test is XLA. Otherwise, I think it's good to go! Could you take a look at it?

**Pull request closed without merging**